### PR TITLE
feat: grpc pipeline support

### LINKED ISSUE #439: replicator pipeline support on grpc impl
### Your question

Now we support replicator pipeline on the bolt impl, but not on the GRPC impl.

### Environment

- SOFAJRaft version:
- JVM version (e.g. `java -version`):
- OS version (e.g. `uname -a`):
- Maven version:
- IDE version:

### PR DESCRIPTION
**Pull request by fengjiachun**
### Motivation:

grpc pipeline support

### Modification:
 1. Attach jraft `Connection` to netty channel (ConnectionInterceptor), for connection closed event and setting attributes.
 2. Use grpc's directExecutor to avoid thread switch (`client.directExecutor()/server. directExecutor()`)
 3. Parse `AppendEntriesRequest` to `AppendEntriesRequestHeader` used for `ExecutorSelector`,  let the specified thread handle the corresponding tasks according to the sequence.

### Result:

Fixes #439

### CODE REVIEW

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java [hunk 1]
```diff
@@ -46,26 +60,48 @@
...
         }
+
+        this.defaultExecutor = ThreadPoolUtil.newBuilder() //
+            .poolName("grpc-default-executor") //
+            .enableMetric(true) //
+            .coreThreads(Math.min(20, GrpcRaftRpcFactory.RPC_SERVER_PROCESSOR_POOL_SIZE / 5)) //
+            .maximumThreads(GrpcRaftRpcFactory.RPC_SERVER_PROCESSOR_POOL_SIZE) //
+            .keepAliveSeconds(60L) //
+            .workQueue(new SynchronousQueue<>()) //
+            .threadFactory(new NamedThreadFactory("grpc-default-executor-", true)) //
```
Comment (zongtanghu) [inline jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java:97]: Herer, it should change to "grpc-default-executor"?

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java [hunk 2]
```diff
@@ -46,26 +60,48 @@
...
     }
 
     @Override
     public boolean init(final Void opts) {
         if (!this.started.compareAndSet(false, true)) {
             throw new IllegalStateException("grpc server has started");
         }
+
+        this.defaultExecutor = ThreadPoolUtil.newBuilder() //
+            .poolName("grpc-default-executor") //
```
Comment (zongtanghu) [inline jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java:91]: String variabler "grpc-default-executor" comes out in serveral places, you can define a private final string variable.

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java [hunk 3]
```diff
@@ -103,48 +140,93 @@ public void registerProcessor(final RpcProcessor processor) {
...
                         @Override
                         public String getRemoteAddress() {
+                            // Rely on GRPC's capabilities, not magic (netty channel)
                             return remoteAddress != null ? remoteAddress.toString() : null;
                         }
                     };
 
-                    processor.handleRequest(rpcCtx, request);
+                    final RpcProcessor.ExecutorSelector selector = processor.executorSelector();
+                    Executor executor;
```
Comment (zongtanghu) [inline jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java:174]: Here, executor should be initilize to null?

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java (hunk 1 above)
Comment (fengjiachun) [reply jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java:97]: its's a prefix, need '-'

File: jraft-extension/rpc-grpc-impl/src/main/java/io/grpc/netty/shaded/io/grpc/netty/NettyConnectionHelper.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    private static Connection attachChannel(final Channel channel, final List<ConnectionClosedEventListener> listeners) {
+        if (channel == null) {
+            return null;
+        }
+
+        final Attribute<NettyConnection> attr = channel.attr(NETTY_CONN_KEY);
+        NettyConnection conn = attr.get();
+        if (conn == null) {
+            final NettyConnection newConn = new NettyConnection(channel);
+            conn = attr.setIfAbsent(newConn);
```
Comment (zongtanghu) [inline jraft-extension/rpc-grpc-impl/src/main/java/io/grpc/netty/shaded/io/grpc/netty/NettyConnectionHelper.java:63]: The connection is thread safe?

Comment (fengjiachun) [reply jraft-extension/rpc-grpc-impl/src/main/java/io/grpc/netty/shaded/io/grpc/netty/NettyConnectionHelper.java:63]: safe

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/ConnectionInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.internal.ServerStream;
+import io.grpc.internal.ServerStreamHelper;
+import io.grpc.netty.shaded.io.grpc.netty.NettyConnectionHelper;
+
+import com.alipay.sofa.jraft.rpc.Connection;
+
+/**
+ *
```
Comment (killme2008) [inline jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/ConnectionInterceptor.java:34]: Adds class doc plz

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcClient.java [hunk 6]
```diff
@@ -107,12 +100,28 @@ public void registerConnectEventListener(final ReplicatorGroup replicatorGroup)
...
+            public Executor executor() {
+                return null;
+            }
+        }, timeoutMs);
+
         try {
-            return ClientCalls.blockingUnaryCall(ch, method, CallOptions.DEFAULT, (Message) request);
+            return future.get(timeoutMs, TimeUnit.MILLISECONDS);
+        } catch (final TimeoutException e) {
+            throw new InvokeTimeoutException(e);
```
Comment (killme2008) [inline jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcClient.java:124]: call `future.cancel(true)` when throwing exception ?

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java [hunk 7]
```diff
@@ -46,26 +60,49 @@
...
+        registerDefaultServerInterceptor();
     }
 
     @Override
     public boolean init(final Void opts) {
         if (!this.started.compareAndSet(false, true)) {
             throw new IllegalStateException("grpc server has started");
         }
+
+        final String name = "grpc-default-executor";
```
Comment (killme2008) [inline jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java:90]: `name` should be constant

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java [hunk 8]
```diff
@@ -103,48 +141,93 @@ public void registerProcessor(final RpcProcessor processor) {
...
+                            } catch (final Throwable t) {
+                                LOG.warn("[GRPC] failed to send response: {}.", t);
+                            }
                         }
 
                         @Override
                         public Connection getConnection() {
-                            throw new UnsupportedOperationException("unsupported");
+                            if (conn == null) {
+                                throw new UnsupportedOperationException("unsupported");
```
Comment (killme2008) [inline jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java:162]: UnsupportedOperationException?   Maybe `IllegalStateException` is more proble.

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java (hunk 7 above)
Comment (fengjiachun) [reply jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java]: done

File: jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java (hunk 8 above)
Comment (fengjiachun) [reply jraft-extension/rpc-grpc-impl/src/main/java/com/alipay/sofa/jraft/rpc/impl/GrpcServer.java]: done

**Pull request merged**